src/databricks/labs/ucx/progress/history.py

```diff
@@ -0,0 +1,216 @@
+        return [getattr(record, field) for field in self._id_attribute_names]
+
+    @classmethod
+    def _encode_non_serializable(cls, name: str, value: Any) -> Any:
+        if isinstance(value, dt.datetime):
+            # Only allow tz-aware timestamps.
+            if value.tzinfo is None:
+                # Name refers to the outermost field, not necessarily a field on a (nested) dataclass.
+                msg = f"Timestamp without timezone not supported in or within field {name}: {value}"
+                raise ValueError(msg)
+            # Always store with 'Z'.
+            ts_utc = value.astimezone(dt.timezone.utc)
+            return ts_utc.isoformat().replace("+00:00", "Z")
+
+        msg = f"Cannot encode {type(value)} value in or within field {name}: {value!r}"
+        raise TypeError(msg)
+
+    def _encode_field_value(self, name: str, value: Any | None) -> str | None:
+        if value is None:
+            return None
+        value_type = self._field_types[name]
+        if value_type in (str, (str | None)):
+            return value
+        encoded_value = json.dumps(
+            value,
+            allow_nan=False,
+            separators=(",", ":"),
+            default=lambda o: self._encode_non_serializable(name, o),
+        )
+        # Handle encoding substituted values that encode as just a string (eg. timestamps); we just return the string.
+        return json.loads(encoded_value) if encoded_value.startswith('"') else encoded_value
+
+    def _object_data_and_failures(self, record: Record) -> tuple[dict[str, str], list[str]]:
+        record_values = self._as_dict(record)
+        encoded_fields = {field: self._encode_field_value(field, record_values[field]) for field in self._field_types}
+        # We must return a value: strings are mandatory (not optional) as the type. As such, optional fields need to be
+        # omitted from the data map if the value is None.
+        data = {k: v for k, v in encoded_fields.items() if v is not None}
+        if self._has_failures == list[str]:
+            failures = record_values["failures"]
+        elif self._has_failures == str:
+            raw_failures = record_values["failures"]
+            try:
+                failures = json.loads(raw_failures) if raw_failures else []
+            except json.decoder.JSONDecodeError:
+                failures = [raw_failures]
+        else:
+            failures = []
+
+        return data, failures
+
+    def to_historical(self, record: Record) -> Historical:
+        data, failures = self._object_data_and_failures(record)
+        return Historical(
+            workspace_id=self._workspace_id,
+            job_run_id=self._job_run_id,
+            object_type=self._object_type,
+            object_id=self._object_id(record),
+            data=data,
+            failures=failures,
+            owner=self._ownership.owner_of(record),
+        )
+
+
+class HistoryLog(Generic[Record]):
```

Comment:
A class is typically named as a persona:

```suggestion
class HistoryLogger(Generic[Record]):
```

Note: I am not sure about the `Logger` as it sounds like Python logging stuff. `Recorder` is maybe more accurate